src/data/device_dmatrix.cu

```diff
@@ -0,0 +1,251 @@
+/*!
+ * Copyright 2020 by Contributors
+ * \file device_dmatrix.cu
+ * \brief Device-memory version of DMatrix.
+ */
+
+#include <thrust/execution_policy.h>
+#include <thrust/iterator/discard_iterator.h>
+#include <thrust/iterator/transform_output_iterator.h>
+#include <xgboost/base.h>
+#include <xgboost/data.h>
+#include <memory>
+#include <utility>
+#include "../common/hist_util.h"
+#include "../common/math.h"
+#include "adapter.h"
+#include "device_adapter.cuh"
+#include "ellpack_page.cuh"
+#include "device_dmatrix.h"
+
+namespace xgboost {
+namespace data {
+
+struct IsValidFunctor : public thrust::unary_function<Entry, bool> {
```

Comment:
Is this copied from `simple_dmatrix.cu`?

Reply:
Yes it's the same but I don't see a convenient place to put it.

Reply:
> Yes it's the same but I don't see a convenient place to put it.

`adapter.cuh`?